nvflare/app_opt/compression/model_compressor.py

```diff
@@ -0,0 +1,128 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Union
+
+import numpy as np
+import torch
+from bitsandbytes.functional import quantize_blockwise
+
+from nvflare.apis.dxo import DXO, DataKind, MetaKey
+from nvflare.apis.dxo_filter import DXOFilter
+from nvflare.apis.fl_context import FLContext
+from nvflare.apis.shareable import Shareable
+from nvflare.app_opt.compression.constant import COMPRESSION_TYPE, DATA_TYPE
+
+
+class ModelCompressor(DXOFilter):
+    def __init__(
+        self,
+        source_data_type="float32",
+        compression_type="float16",
+    ):
+        """Filter to compress Shareable object to reduce communication burden.
+
+        Args:
+            source_data_type: original data type of the model
+            compression_type: method used for compression
+
+        """
+
+        # support weight and weight_diff data kinds
+        data_kinds = [DataKind.WEIGHTS, DataKind.WEIGHT_DIFF]
+        super().__init__(supported_data_kinds=data_kinds, data_kinds_to_filter=data_kinds)
+
+        # assign data and compression types
+        self.logger.info("Using model compressor.")
+        # check if source data type is valid
+        if source_data_type.upper() not in DATA_TYPE:
+            raise ValueError(f"Invalid source data type: {source_data_type}")
+        else:
+            self.source_data_type = source_data_type
+        # check if compression type is valid
+        if compression_type.upper() not in COMPRESSION_TYPE:
+            raise ValueError(f"Invalid compression type: {compression_type}")
+        else:
+            self.compression_type = compression_type
+        # compression constants
+        self.FP16_MIN = np.finfo(np.float32).min
```

Comment:
Shouldn't this depend on `source_data_type`?

Reply:
yes, initiate here to avoid multiple finfo calling